src/ServiceControl/Operations/FailedMessagePersister.cs

```diff
@@ -46,28 +47,31 @@ public async Task<FailureDetails> Persist(MessageContext message)
                 messageId = DeterministicGuid.MakeId(message.MessageId).ToString();
             }
 
-            var metadata = new Dictionary<string, object>
+            var metadata = new ConcurrentDictionary<string, object>
             {
                 ["MessageId"] = messageId,
                 ["MessageIntent"] = message.Headers.MessageIntent(),
                 ["HeadersForSearching"] = string.Join(" ", message.Headers.Values)
             };
 
-            // TODO: fanout ? 
+            var enricherTasks = new List<Task>(enrichers.Length);
+            // ReSharper disable once LoopCanBeConvertedToQuery
             foreach (var enricher in enrichers)
             {
-                await enricher.Enrich(message.Headers, metadata)
-                    .ConfigureAwait(false);
+                enricherTasks.Add(enricher.Enrich(message.Headers, metadata));
             }
 
+            await Task.WhenAll(enricherTasks)
+                .ConfigureAwait(false);
+
             await bodyStorageEnricher.StoreErrorMessageBody(message.Body, message.Headers, metadata)
                 .ConfigureAwait(false);
 
             var failureDetails = failedMessageFactory.ParseFailureDetails(message.Headers);
 
             var processingAttempt = failedMessageFactory.CreateProcessingAttempt(
                 message.Headers,
-                metadata,
+                new Dictionary<string, object>(metadata), 
```

Comment:
This would be the consequence but I think this is an OK price to pay given that we can fan out the enrichers which do domain event raise and other IO bound stuff?